v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/DataSourceProvider.java

```diff
@@ -0,0 +1,103 @@
+ * the License.
+ */
+package com.google.cloud.teleport.v2.source;
+
+import static com.google.cloud.teleport.v2.utils.KMSUtils.maybeDecrypt;
+
+import com.google.cloud.teleport.v2.options.SourceDbToSpannerOptions;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.apache.beam.sdk.io.jdbc.JdbcIO;
+import org.apache.beam.sdk.options.ValueProvider.StaticValueProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DataSourceProvider
+    implements org.apache.beam.sdk.transforms.SerializableFunction<Void, javax.sql.DataSource> {
+
+  private static final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+  private static final Lock writeLock = readWriteLock.writeLock();
+  private static final Lock readLock = readWriteLock.readLock();
+  private static final Logger LOG = LoggerFactory.getLogger(DataSourceProvider.class);
+  private static volatile javax.sql.DataSource dataSource = null;
+  private final JdbcIO.DataSourceConfiguration config;
+
+  public DataSourceProvider(SourceDbToSpannerOptions options) {
+    config = getDataSourceConfiguration(options);
+  }
+
+  private static JdbcIO.DataSourceConfiguration getDataSourceConfiguration(
+      SourceDbToSpannerOptions options) {
+    var config =
+        JdbcIO.DataSourceConfiguration.create(
+                StaticValueProvider.of(options.getJdbcDriverClassName()),
+                maybeDecrypt(options.getSourceConnectionURL(), null))
+            .withUsername(maybeDecrypt(options.getUsername(), null))
+            .withPassword(maybeDecrypt(options.getPassword(), null))
+            .withMaxConnections(options.getMaxConnections());
+
+    if (options.getSourceConnectionProperties() != null) {
+      config = config.withConnectionProperties(options.getSourceConnectionProperties());
+    }
+    if (options.getJdbcDriverJars() != null) {
+      config = config.withDriverJars(options.getJdbcDriverJars());
+    }
+    return config;
+  }
+
+  private javax.sql.DataSource initializeDataSourceIfRequired() {
+    // Initialize the connection pool if required.
+    javax.sql.DataSource ret = null;
+    ret = getDataSource();
+    if (ret == null) {
+      ret = initializeDataSource();
+    }
+    return ret;
+  }
+
+  private javax.sql.DataSource initializeDataSource() {
+    javax.sql.DataSource ret;
+    writeLock.lock();
+    try {
+      // Check if it was already initialized till the write lock was acquired.
+      if (dataSource == null) {
+        dataSource = JdbcIO.PoolableDataSourceProvider.of(config).apply(null);
+        LOG.debug("initializeDataSource dataSource" + dataSource.toString());
+      }
+      ret = dataSource;
+    } finally {
+      writeLock.unlock();
+    }
+    return ret;
+  }
+
+  private javax.sql.DataSource getDataSource() {
+    javax.sql.DataSource ret = null;
+    readLock.lock();
```

Comment:
why do we need a read lock here ?

Reply:
The read-write locking was used to achivee both - synchronization and order the double checked locking loads/stores. Thanks for pointing me to double-checked locking in the review. 
I have implemented the recommended (or idiomatic for java9+) double checked lock in this push.